### ISSUE TITLE
storage: TestTransferRaftLeadership is flaky

### ISSUE THREAD

**Issue by d4l3k** [cockroachdb/cockroach#7666]
Found with  stress on master 98b248e3981d6a8b16fac061fb8e004e7eed0c50
```
make stress PKG=./storage STRESSFLAGS="-stderr -maxfails 1 -p=16" TESTS="TestTransferRaftLeadership"
```

```
cd ./storage && stress -stderr -maxfails 1 -p=16 /Users/wiz/go/src/github.com/cockroachdb/cockroach/stress.test -test.run "TestTransferRaftLeadership" -test.timeout 2m

I160706 17:42:01.611461 rand.go:76  Random seed: 5608689081079303159
I160706 17:42:01.613959 storage/engine/rocksdb.go:289  opening in memory rocksdb instance
W160706 17:42:01.615427 gossip/gossip.go:900  not connected to cluster; use --join to specify a connected node
W160706 17:42:01.617890 gossip/gossip.go:900  not connected to cluster; use --join to specify a connected node
I160706 17:42:01.619017 storage/replica_command.go:1576  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=1970-01-01 00:00:00 +0000 UTC]
I160706 17:42:01.619432 storage/engine/rocksdb.go:289  opening in memory rocksdb instance
W160706 17:42:01.620406 gossip/gossip.go:900  not connected to cluster; use --join to specify a connected node
I160706 17:42:01.621724 storage/engine/rocksdb.go:289  opening in memory rocksdb instance
I160706 17:42:01.622274 gossip/gossip.go:926  starting client to 127.0.0.1:61326
W160706 17:42:01.822224 gossip/gossip.go:900  not connected to cluster; use --join to specify a connected node
I160706 17:42:01.840545 gossip/gossip.go:926  starting client to 127.0.0.1:61326
I160706 17:42:01.842302 storage/replica_raftstorage.go:442  generated snapshot for range 1 at index 12 in 180.124µs. encoded size=2631, 28 KV pairs, 2 log entries
I160706 17:42:01.842612 storage/replica_command.go:2707  change replicas of 1: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2
I160706 17:42:01.843261 storage/replica.go:1911  range 1: applied part of replica change txn: BeginTransaction [/Local/Range/""/RangeDescriptor,/Min), ConditionalPut [/Local/Range/""/RangeDescriptor,/Min), Put [/Meta2/Max,/Min), Put [/Meta1/Max,/Min), pErr=<nil>
I160706 17:42:01.843680 storage/replica.go:1384  raft: proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I160706 17:42:01.844038 storage/replica.go:1911  range 1: applied part of replica change txn: EndTransaction [/Local/Range/""/RangeDescriptor,/Min), pErr=<nil>
I160706 17:42:01.844785 storage/replica_raftstorage.go:442  generated snapshot for range 1 at index 14 in 282.414µs. encoded size=3482, 31 KV pairs, 4 log entries
I160706 17:42:01.845080 storage/replica_command.go:2707  change replicas of 1: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3
I160706 17:42:01.845782 storage/replica_raftstorage.go:548  replica 0 received snapshot for range 1 at index 12. encoded size=2631, 28 KV pairs, 2 log entries
I160706 17:42:01.846328 storage/replica_raftstorage.go:551  replica 0 applied snapshot for range 1 in 509.589µs
I160706 17:42:01.846338 storage/replica_raftstorage.go:548  replica 0 received snapshot for range 1 at index 14. encoded size=3482, 31 KV pairs, 4 log entries
I160706 17:42:01.846802 storage/replica_raftstorage.go:551  replica 0 applied snapshot for range 1 in 424.634µs
I160706 17:42:01.848313 storage/replica.go:1911  range 1: applied part of replica change txn: BeginTransaction [/Local/Range/""/RangeDescriptor,/Min), ConditionalPut [/Local/Range/""/RangeDescriptor,/Min), Put [/Meta2/Max,/Min), Put [/Meta1/Max,/Min), pErr=<nil>
I160706 17:42:01.848742 storage/replica.go:1384  raft: proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I160706 17:42:01.903559 storage/replica.go:1911  range 1: applied part of replica change txn: BeginTransaction [/Local/Range/""/RangeDescriptor,/Min), ConditionalPut [/Local/Range/""/RangeDescriptor,/Min), Put [/Meta2/Max,/Min), Put [/Meta1/Max,/Min), pErr=<nil>
W160706 17:42:01.918680 raft/raft.go:593  [range 1] 1 stepped down to follower since quorum is not active
I160706 17:42:01.918746 storage/replica.go:1384  raft: proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I160706 17:42:01.918866 storage/replica.go:1911  range 1: applied part of replica change txn: EndTransaction [/Local/Range/""/RangeDescriptor,/Min), pErr=<nil>
I160706 17:42:02.113979 storage/replica.go:1911  range 1: applied part of replica change txn: BeginTransaction [/Local/Range/""/RangeDescriptor,/Min), ConditionalPut [/Local/Range/""/RangeDescriptor,/Min), Put [/Meta2/Max,/Min), Put [/Meta1/Max,/Min), pErr=<nil>
I160706 17:42:02.221646 storage/replica.go:1384  raft: proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I160706 17:42:02.225653 storage/replica.go:1911  range 1: applied part of replica change txn: EndTransaction [/Local/Range/""/RangeDescriptor,/Min), pErr=<nil>
I160706 17:42:02.226406 storage/replica.go:1911  range 1: applied part of replica change txn: EndTransaction [/Local/Range/""/RangeDescriptor,/Min), pErr=<nil>
I160706 17:42:02.226876 stopper.go:355  draining; tasks left:
1      kv/txn_coord_sender.go:870
W160706 17:42:02.226984 storage/replica.go:326  range 1: outgoing raft transport stream to {3 3 0} closed by the remote: rpc error: code = 13 desc = transport is closing
W160706 17:42:02.227303 storage/replica.go:326  range 1: outgoing raft transport stream to {1 1 1} closed by the remote: rpc error: code = 13 desc = transport is closing
W160706 17:42:02.227352 storage/replica.go:326  range 1: outgoing raft transport stream to {2 2 2} closed by the remote: rpc error: code = 13 desc = transport is closing
W160706 17:42:02.227373 storage/replica.go:326  range 1: outgoing raft transport stream to {2 2 0} closed by the remote: rpc error: code = 13 desc = transport is closing
--- FAIL: TestTransferRaftLeadership (0.62s)
	client_raft_test.go:1786: raft leader should be 1, but got 2
FAIL


ERROR: exit status 1

61 runs completed, 1 failures, over 2s
FAIL
```

Also occurred in https://circleci.com/gh/cockroachdb/cockroach/19948 with which may be unrelated.
```
	client_raft_test.go:1794: expect get nil, actual get range 1 was not found 
```

**Comment by bdarnell** [comment]
Maybe be related to the issue being fixed upstream in coreos/etcd#5809

**Comment by xiang90** [comment]
The upstream issue is fixed.

**Comment by petermattis** [comment]
@d4l3k Is our test fixed now that the upstream issue is fixed?

**Comment by d4l3k** [comment]
Seems to be fixed. Not seeing any flakes when stressing it.

**Comment by bdarnell** [comment]
This is still slightly flaky. The upstream issue has been fixed but we haven't updated our dependency yet because we're blocked on other issues (#8017). It reproduces in seconds locally with `make stressrace PKG=./storage TESTS=TransferRaftLeadership STRESSFLAGS='-p 16 -maxfails 1 -stderr'` (but not at all with the default `-p` value). (I saw it in a [PR](https://circleci.com/gh/cockroachdb/cockroach/21245) that changed the raft transport, but it's flaky on master too)

Looking closer, the failures I'm seeing now are different from the one quoted above:
```
--- FAIL: TestTransferRaftLeadership (0.90s)
	client_raft_test.go:2121: raft leader should be 1, but got status {"id":"1","term":7,"vote":"2","commit":18,"lead":"2","raftState":"StateFollower","progress":{}}
FAIL
```

This is earlier in the test, when it's establishing initial conditions before it's even tried the transfer.
